dhall-lsp-server/src/Dhall/LSP/Backend/Completion.hs

```diff
@@ -0,0 +1,186 @@
+        context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+        values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e
+
+  -- if nothing works, only remember the name (but bind to `holeExpr`)
+  | otherwise =
+    let context' = fmap (shift 1 (V x 0)) $ insert x holeExpr context
+        values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' e
+
+buildCompletionContext' context values (Lam x _A b) =
+  let _A' | Right _ <- typeWithA absurd context _A = normalize _A
+          | otherwise = holeExpr
+
+      context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+      values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' b
+
+buildCompletionContext' context values (Pi x _A b) =
+  let _A' | Right _ <- typeWithA absurd context _A = normalize _A
+          | otherwise = holeExpr
+
+      context' = fmap (shift 1 (V x 0)) $ insert x _A' context
+      values' = fmap (shift 1 (V x 0)) $ insert x holeExpr values
+
+    in buildCompletionContext' context' values' b
+
+-- catch-all
+buildCompletionContext' context values _ = CompletionContext context values
+
+-- Helper. Given `Dhall.Context.toList ctx` construct the corresponding variable
+-- names.
+contextToVariables :: [(Text, Expr Src X)] -> [Var]
+contextToVariables  [] = []
+contextToVariables ((name, _) : rest) =
+  V name 0 : map (inc name) (contextToVariables rest)
+ where inc x (V y i) | x == y = V x (i + 1)
+                     | otherwise = V y i
+
+-- | Complete identifiers from the given completion context.
+completeFromContext :: CompletionContext -> [Completion]
+completeFromContext (CompletionContext context _) =
+  let context' = toList context
+      completeReserved keyword
+        | Right expr <- exprFromText "" keyword
+        , Right typ <- typeOf (fmap undefined expr) =
+          Completion keyword (Just typ)
+        | otherwise = Completion keyword Nothing
+      reserved = map completeReserved $ HashSet.toList reservedIdentifiers
+  in [ Completion (pretty var) (if typ == holeExpr then Nothing else Just typ)
+     | (var, (_, typ)) <- zip (contextToVariables context') (toList context) ]
+     ++ reserved
+
+-- | Complete union constructors and record projections.
+completeProjections :: CompletionContext -> Expr Src X -> [Completion]
+completeProjections (CompletionContext context values) expr =
+  -- substitute 'dependent lets', necessary for completion of unions
+  let values' = toList values
+      subs = filter ((/= holeExpr) . snd) $ zip (contextToVariables values') (map snd values')
+      expr' = foldl (\e (x,val) -> subst x val e) expr subs
+
+  in case typeWithA absurd (traceShow (map fst $ toList context) $ context) expr' of
+      Left _ -> []
+      Right _A ->
+        let expr'' = normalize expr'
+        in completeUnion expr'' expr'' ++ completeRecord (normalize _A)
```

Comment:
I think `completeUnion` only needs to take `expr''` as an argument once, if you change it to:

```
completeUnion _A@(Union m) = ...
```

Reply:
No, I'm using the copy of the initial argument to complete the resulting type while recursing on the other, e.g. `< Left | Right>.Left` has type `<Left | Right>`.